### PR TITLE
Render documentation for provider `init` callbacks

### LINKED ISSUE #182: Provider `init` callback needs to be documented
If a provider has `init` specified, we want to document it.

In other words,

```bzl
def _my_info_init(a, c = "hello"):
    """
    MyInfo constructor

    Args:
          a: Value of field a
          c: Value to frobnicate
    """
    return {"a": a, "b": frobnicate(c)}

MyInfo = provider(
    doc = """This is my provider""",
    fields = {
        "a": "Field a",
        "b": "Field b (frobnicated)",
    },
    init = _my_info_init,
)
```

should get documented as something like

----

## MyInfo

This is my provider

<pre>
MyInfo(<a href="#MyInfo-_init-a">a</a>, <a href="##MyInfo-_init-c">c</a>)
</pre>

MyInfo constructor

**CONSTRUCTOR PARAMETERS**

 Name  | Description | Default Value |
| :------------- | :------------- | :------------- |
| <a id="MyInfo-_init-a"></a>a |  Value of field a   |  none |
| <a id="MyInfo-_init-c"></a>c |  Value to frobnicate  |  `"hello"` |

**FIELDS**

| Name  | Description |
| :------------- | :------------- |
| <a id="MyInfo-a"></a>a |  Field a  |
| <a id="MyInfo-b"></a>b |  Field b (frobnicated)  |

----

This is a prerequisite for switching to stardoc / starlark_doc_extract for generating Bazel's own documentation.

### PR DESCRIPTION
**Pull request by tetromino**
If a provider has a custom `init` callback, we want the summary blurb to show `init`'s parameters (since these are what the user will interact with); we have to render constructor paramaters and fields separately (and using separate html anchors) since there is not necessarily a 1-1 relationship between them, and since they may have different docs.

Fixes #182

### CODE REVIEW

**Comment by tetromino** [comment]
Current PR is not ready. I need to fix the following problems which occurred to me in the process of testing it:

*  In virtually all cases, the `init` callback's summary docstring is *not interesting to the user* (it's along the lines of "Validates params foo and bar") - so we should not render it. The provider's docstring is the interesting part for the user.
* Typically, the list of parameters of the `init` callback exactly matches the list of fields. If that is so, we should merge parameters and fields into a single table.
* And even if the set of `init` parameters is different from fields, we should propagate docstrings from fields to undocumented init parameters with the same name.

In other words, fairly extensive munging of the ProviderInfo proto is necessary.

**Comment by tetromino** [comment]
I think by default, we want the following behavior:

* Custom init callback specified
  * The set of parameters for the init callback equals the set of
    fields for the provider; and the docs for the init callback's
    parameters are either empty or equal to corresponding field docs
    * Some init parameters have a default value:
      -> Render a single "Fields" table with 3 columns (name, doc,
         default value)
    * ... otherwise
      -> Render a single "Fields" table with 2 columns
  * ... otherwise
    -> Render two tables - "Constructor parameters" and "Fields" - with
       the links from the summary blurb (interfixed with "_init")
       leading to the parameters table (not the fields table)
* ... otherwise
  -> Trivial case - single "Fields" table (as before).

**Review by brandjon — APPROVED** [review]
Discussed offline, approved but mind the rendering failure in one of the goldens,
